Conditional content changes

cselvaraj opened the pull request:
ANS changes needed for Conditional Content, added to the current ANS dev version.

PSTK-2743: As a developer, I want to understand the interface and data structures that will be required for the proposed solution
https://arcpublishing.atlassian.net/browse/PSTK-2743

cselvaraj committed:
New version prep

PSTK-2743: As a developer, I want to understand the interface and data structures that will be required for the proposed solution
https://arcpublishing.atlassian.net/browse/PSTK-2743

cselvaraj committed:
Conditional content changes

PSTK-2743: As a developer, I want to understand the interface and data structures that will be required for the proposed solution
https://arcpublishing.atlassian.net/browse/PSTK-2743

cselvaraj committed:
Conditional content changes

PSTK-2743: As a developer, I want to understand the interface and data structures that will be required for the proposed solution
https://arcpublishing.atlassian.net/browse/PSTK-2743

cselvaraj committed:
Add a field in variant for the variant content:

- this will be missing in simple variant metadata listing data, but could be included in more comprehensive representations

PSTK-2743: As a developer, I want to understand the interface and data structures that will be required for the proposed solution
https://arcpublishing.atlassian.net/browse/PSTK-2743

cselvaraj committed:
Add a field in variant for the variant content:

- this will be missing in simple variant metadata listing data, but could be included in more comprehensive representations

PSTK-2743: As a developer, I want to understand the interface and data structures that will be required for the proposed solution
https://arcpublishing.atlassian.net/browse/PSTK-2743

hartmanc-wp commented on src/main/resources/schema/ans/0.10.10/utils/variant.json:
```diff
@@ -0,0 +1,32 @@
+    "additional_properties": {
+      "$ref": "https://raw.githubusercontent.com/washingtonpost/ans-schema/master/src/main/resources/schema/ans/0.10.10/traits/trait_additional_properties.json"
+    },
+    "content": {
```
Can you give an example of what this will look like as JSON data? I'm unclear how this will look in practice.

cselvaraj replied:
In a general sense, it would be the content type that has variant data.  The immediate case is for story data, so this is an example:

```
{
  "content_elements": [
    {
      "_id": "some_placeholder_ID",
      "content_elements": [
        {
          "content": "Some content zone data",
          "type": "text"
        }
      ],
      "type": "element_group"
    }
  ],
  "promo_items": {
    "basic": {
      "referent": {
        "id": "some image ID",
        "type": "image"
      },
      "type": "reference"
    }
  },
  "taxonomy": {
    "tags": [
      {
        "slug": "some-local-tag",
        "text": "some local tag",
        "type": "tag"
      }
    ]
  },
  "type": "story",
  "version": "0.10.8"
}
```

The logic to assemble the final story would read this story for the overrides and content zones data.

hartmanc-wp replied:
Makes sense.  Is there a way to enforce the structure of content with ANS?

manginav replied:
yes the `required` fields enforces the schema for ANS.

manginav replied:
Also @cselvaraj  and @hartmanc-wp  I think we need to follow the way `story` had the schema [here](https://github.com/washingtonpost/ans-schema/blob/master/src/main/resources/schema/ans/0.10.9/story.json), so that we exclusively mention what are the fields we want in our variant. 

1. content [here](https://raw.githubusercontent.com/washingtonpost/ans-schema/master/src/main/resources/schema/ans/0.10.9/content.json) does not have `content_elements` 
2. And in content there are lot of fields which we are not using [ not a real problem though ]

cselvaraj replied:
The choice to use the "content" trait was to allow this field to hold different, top-level types.  Instead of it, we could use the JSON schema `oneOf` directive and explicitly list the specific content types that we allow.  That would force us to update when we want to extend variants to other types, but it has the benefit of being more explicit.

manginav replied:
hey @cselvaraj to the example above you mentioned there are content_elements where in the content object [schema](https://raw.githubusercontent.com/washingtonpost/ans-schema/master/src/main/resources/schema/ans/0.10.9/content.json) there is no such field, so is it because we dont want to enforce content_elements in variant schema?

hartmanc-wp commented on src/main/resources/schema/ans/0.10.10/utils/variant.json:
```diff
@@ -0,0 +1,32 @@
+      "$ref": "https://raw.githubusercontent.com/washingtonpost/ans-schema/master/src/main/resources/schema/ans/0.10.10/content.json",
+      "description": "Variant content"
+    },
+    "websites": {
```
What about putting this into a "conditions" field, which contains a "website_condition" object.  That way it could get extended pretty easily to future conditions.

cselvaraj replied:
I erred on the side of keeping things simple.  We don't really have a strong idea of what future conditions will look like, so the "websites" field will satisfy the current requirements.  We can expand later by using the JSON schema's "[oneOf](https://json-schema.org/draft/2020-12/json-schema-core.html#rfc.section.10.2.1.3)" keyword to later replace the use of the "websites" field.  This gives us simplicity now, but we'll be able to change things later, while keeping backwards compatibility.

hartmanc-wp commented on src/main/resources/schema/ans/0.10.10/traits/trait_variants.json:
```diff
@@ -0,0 +1,28 @@
+    "additional_properties": {
+      "$ref": "https://raw.githubusercontent.com/washingtonpost/ans-schema/master/src/main/resources/schema/ans/0.10.10/traits/trait_additional_properties.json"
+    },
+    "placeholder_ids": {
```
What's the value added here?  It's helpful, but also redundant since you have them in the "content"elements" of the hub story.

cselvaraj replied:
Those won't be present until the variant content exists.  This defines them first so they can be used when the content is created.  Additionally, the variant content isn't guaranteed to override all of the zones, so this just defined the zones that are possible.

manginav replied:
makes sense.

hartmanc-wp commented on src/main/resources/schema/ans/0.10.10/story.json:
```diff
@@ -0,0 +1,181 @@
+    "publishing": {
+      "$ref": "https://raw.githubusercontent.com/washingtonpost/ans-schema/master/src/main/resources/schema/ans/0.10.10/traits/trait_publishing.json"
+    },
+    "variants": {
```
Doesn't this imply that all variants and their data will be inflated here into the hub story for content api?

cselvaraj replied:
The idea is to mimic the "websites" field for stories.  This field will give metadata for the variants without giving all the variant content data.  The schema does support the full data, but it's an optional field, so does not need to be present.  That property may be useful in certain circumstances (e.g., database representations,) so the possibility was offered, but it probably won't be utilized in most cases.

hartmanc-wp commented on src/main/resources/schema/ans/0.10.10/utils/variant.json:
```diff
@@ -0,0 +1,32 @@
+  "required": [
+    "_id"
+  ],
+  "title": "Variant metadata",
```
Where are we going to return the `published` and the timestamp fields?

cselvaraj replied:
Good call.  That was an oversight.  I'll add those fields.

cselvaraj committed:
PR feedback

PSTK-2743: As a developer, I want to understand the interface and data structures that will be required for the proposed solution
https://arcpublishing.atlassian.net/browse/PSTK-2743

cselvaraj committed:
PR feedback

PSTK-2743: As a developer, I want to understand the interface and data structures that will be required for the proposed solution
https://arcpublishing.atlassian.net/browse/PSTK-2743

cselvaraj committed:
Rename "placeholder" to "content_zone."

PSTK-2743: As a developer, I want to understand the interface and data structures that will be required for the proposed solution
https://arcpublishing.atlassian.net/browse/PSTK-2743

cselvaraj committed:
Rename "placeholder" to "content_zone."

PSTK-2743: As a developer, I want to understand the interface and data structures that will be required for the proposed solution
https://arcpublishing.atlassian.net/browse/PSTK-2743

cselvaraj committed:
Make story data the only supported data for variant content.

PSTK-2743: As a developer, I want to understand the interface and data structures that will be required for the proposed solution
https://arcpublishing.atlassian.net/browse/PSTK-2743

cselvaraj committed:
Make story data the only supported data for variant content.

PSTK-2743: As a developer, I want to understand the interface and data structures that will be required for the proposed solution
https://arcpublishing.atlassian.net/browse/PSTK-2743

cselvaraj committed:
Merge with 0.10.9 redo:

- Conditional Content changes will be added to the current dev version

cselvaraj commented:
I redid these changes to integrate them with the current dev ANS version (0.10.9.)  Those are changes that corrected some schema files to make them compliant with the Draft 4 standard.

hartmanc-wp commented:
LGTM 👍
